session-replay/Sources/DatadogSessionReplay/Processor/Diffing/Diff.swift

```diff
@@ -0,0 +1,231 @@
+/// `oldArray` are guaranteed to have different `id` (same for `newArray`). Elements with the same `id` can appear in both arrays, which
+/// indicates one of two things determined by `newElement.isDifferent(than: oldElement)`:
+/// - either the element was not altered and can be skipped in diff,
+/// - or the element was changed and it should be reflected in `Diff.Update`.
+///
+/// Like original algorithm, our implementation uses 6 passes over both arrays to determine diff
+///
+/// Ref.:
+/// - _"A technique for isolating differences between files"_ Paul Heckel (1978) - https://dl.acm.org/citation.cfm?id=359467
+///
+/// - Parameters:
+///   - oldArray: original array
+///   - newArray: new array
+/// - Returns: `Diff` describing changes from `oldArray` to `newArray`.
+internal func computeDiff<E: Diffable>(oldArray: [E], newArray: [E]) throws -> Diff<E> {
+    var table: [DiffableID: Symbol] = [:]
+    var oa: [Entry] = [] // old array entries
+    var na: [Entry] = [] // new array entries
+
+    // 1st pass
+    // Read `newArray` and store info on each element in symbols `table`:
+    for element in newArray {
+        table[element.id] = Symbol(inOld: false, inNew: true, indexInOld: nil)
+        na.append(.reference(element.id))
+    }
+
+    // 2nd pass
+    // Read `oldArray` and store info on each element in symbols `table`. If certain element already
+    // exists, update its information (otherwise create new entry):
+    for (index, element) in oldArray.enumerated() {
+        if table[element.id] == nil {
+            table[element.id] = Symbol(inOld: true, inNew: false, indexInOld: index)
+        } else {
+            table[element.id]?.inOld = true
+            table[element.id]?.indexInOld = index
+        }
+        oa.append(.reference(element.id))
+    }
+
+    // 3rd pass
+    // Uses "Observation 1":
+    // > If a line occurs only once in each file, then it must be the same line, although it may have been moved.
+    // > We use this observation to locate unaltered lines that we subsequently exclude from further treatment.
+    for (index, entry) in na.enumerated() {
+        if case let .reference(id) = entry {
+            guard let symbol = table[id] else {
+                throw DiffError()
+            }
+            if symbol.inOld && symbol.inNew {
+                guard let indexInOld = symbol.indexInOld else {
+                    throw DiffError()
+                }
+                na[index] = .index(indexInOld)
+                oa[indexInOld] = .index(index)
+            }
+        }
+    }
+
+    // 4th pass
+    // > If a line has been found to be unaltered, and the lines immediately adjacent to it in both files are identical,
+    // > then these lines must be the same line. This information can be used to find blocks of unchanged lines.
+    if na.count > 1 {
+        for i in (1..<(na.count - 1)) {
```

Comment:
Why are you starting from 1 here ?

Reply:
Indeed, it looks like the [4th](https://gist.github.com/ndarville/3166060#fourth-pass) and [5th](https://gist.github.com/ndarville/3166060#fifth-pass) pass of the algorithm should process all values of `NA` and `OA`.

Reply:
Yeah, I think it's a leftover after safeguarding boundaries of this iteration. It should start from `0` - it's fixed now 👌.

Interestingly, it wasn't causing any trouble even after I added more fuzzy tests on short and long sequences. For the sake of safety, I pushed enhancement to tests as well 👍.